types/src/network_address.rs

```diff
@@ -0,0 +1,651 @@
+    type Error = lcs::Error;
+
+    fn try_from(value: &NetworkAddress) -> Result<Self, lcs::Error> {
+        let bytes = lcs::to_bytes(value)?;
+        Ok(RawNetworkAddress::new(bytes))
+    }
+}
+
+////////////////////
+// NetworkAddress //
+////////////////////
+
+impl NetworkAddress {
+    fn new(protocols: Vec<Protocol>) -> Self {
+        Self(protocols)
+    }
+}
+
+impl FromStr for NetworkAddress {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.is_empty() {
+            return Err(ParseError::EmptyProtocolString);
+        }
+
+        let mut protocols = Vec::new();
+        let mut parts = s.split('/');
+
+        if parts.next() != Some("") {
+            return Err(ParseError::InvalidProtocolString);
+        }
+
+        while let Some(protocol_type) = parts.next() {
+            let arg = parts.next().ok_or(ParseError::UnexpectedEnd)?;
+            let protocol = match protocol_type {
+                "ip4" => Protocol::Ip4(arg.parse()?),
+                "ip6" => Protocol::Ip6(arg.parse()?),
+                "dns" => Protocol::Dns(arg.parse()?),
+                "dns4" => Protocol::Dns4(arg.parse()?),
+                "dns6" => Protocol::Dns6(arg.parse()?),
+                "tcp" => Protocol::Tcp(arg.parse()?),
+                "memory" => Protocol::Memory(arg.parse()?),
+                "ln-noise-ik" => Protocol::NoiseIk(x25519::PublicKey::from_encoded_string(arg)?),
+                "ln-handshake" => Protocol::Handshake(arg.parse()?),
+                unknown => return Err(ParseError::UnknownProtocolType(unknown.to_string())),
+            };
+            protocols.push(protocol);
+        }
+
+        Ok(NetworkAddress::new(protocols))
+    }
+}
+
+impl TryFrom<Vec<Protocol>> for NetworkAddress {
+    type Error = EmptyError;
+
+    fn try_from(value: Vec<Protocol>) -> Result<Self, Self::Error> {
+        if value.is_empty() {
+            Err(EmptyError)
+        } else {
+            Ok(NetworkAddress::new(value))
+        }
+    }
+}
+
+impl TryFrom<&RawNetworkAddress> for NetworkAddress {
+    type Error = lcs::Error;
+
+    fn try_from(value: &RawNetworkAddress) -> Result<Self, lcs::Error> {
+        let addr: NetworkAddress = lcs::from_bytes(value.as_ref())?;
+        Ok(addr)
+    }
+}
+
+impl fmt::Display for NetworkAddress {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for protocol in self.0.iter() {
+            protocol.fmt(f)?;
+        }
+        Ok(())
+    }
+}
+
+impl fmt::Debug for NetworkAddress {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+impl Serialize for NetworkAddress {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        #[derive(Serialize)]
+        #[serde(rename = "NetworkAddress")]
+        struct SerializeWrapper<'a>(&'a [Protocol]);
```

Comment:
can you add some explanation on why we need to create this `SerializeWrapper` type?

Reply:
right, so I can't just do

```
#[derive(Deserialize, Serialize)]
struct NetworkAddress(Vec<Protocol>);
```

because I need to override the logic for human-readable ser/de. But, I still want to use the generated Deserialize/Serialize for lcs (i.e., not human-readable). So I have a little wrapper in here to get the derive-generated `serialize` in this case.

Reply:
In that case,  should we have `struct SerializeWrapper<'a>(&'a NtworkAddress)`?

Reply:
we're impl'ing custom `Serialize`/`Deserialize` for `NetworkAddress` so that will cause an infinite loop as `serde-derive` for `SerializeWrapper` will generate a `Serialize` that eventually calls this function `NetworkAddress::serialize`. (`SerializeWrapper::serialize` calls `NetworkAddress::serialize` which call's `SerializeWrapper::serialize` ...)